### PR TITLE
Adds an `Update` module used in `EntityOperations`

### PR DESCRIPTION
**Pull request by telser**
This is based on/matches the `Insert` module. To match what we have
for insert/select the Expr/Update module is now split into submodules.
Further some of the haddock headers were wrong for Insert modules so
those are fixed as well.

### CODE REVIEW

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+{- |
+  Extracts the query that will be run when the update is executed. Normally you
+  don't want to extract the query and run it yourself, but this function is
+  useful to view the query for debugging or query explanation.
+-}
+updateToUpdateExpr :: Update readEntity -> Expr.UpdateExpr
+updateToUpdateExpr (Update _ expr) = expr
+
+{- |
+  Excutes the database query for the 'Update' and returns '()'.
```
Comment (jlavelle) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs:37]:
```suggestion
  Executes the database query for the 'Update' and returns '()'.
```

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+-}
+executeUpdate :: MonadOrville.MonadOrville m => Update readEntity -> m ()
+executeUpdate (Update _ expr) =
+  Execute.executeVoid expr
+
+{- | Excutes the database query for the 'Update' and uses its 'SqlMarshaller' to decode the rows (that
```
Comment (jlavelle) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs:43]:
```suggestion
{- | Executes the database query for the 'Update' and uses its 'SqlMarshaller' to decode the rows (that
```

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  were just updateed) as returned via a RETURNING clause.
```
Comment (jlavelle) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs:44]:
```suggestion
  were just updated) as returned via a RETURNING clause.
```

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+-}
+executeUpdateReturnEntities :: MonadOrville.MonadOrville m => Update readEntity -> m (Maybe readEntity)
+executeUpdateReturnEntities (Update marshaller expr) =
+  fmap listToMaybe $ Execute.executeAndDecode expr marshaller
+
+{- |
+  Builds an 'Update' that will update all of the writeable columns described in the
```
Comment (jlavelle) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs:51]:
```suggestion
  Builds an 'Update' that will update all of the writable columns described in the
```

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+updateToTable ::
+  TableDefinition (HasKey key) writeEntity readEntity ->
+  key ->
+  writeEntity ->
+  Update readEntity
+updateToTable =
+  updateTable WithoutReturning
+
+{- |
+  Builds an 'Update' that will update all of the writeable columns described in the
```
Comment (jlavelle) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Update.hs:63]:
```suggestion
  Builds an 'Update' that will update all of the writable columns described in the
```

**Comment by telser** [comment]
@jlavelle Thanks for all those suggestions. I think I got them all, plus a plural that wasn't really plural.

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Insert.hs [hunk 6]
```diff
@@ -89,5 +90,6 @@ insertTable returningOption tableDef entities =
   that Orville supports using the expression building functions, or use @RawSql.fromRawSql@ to build
   a raw 'Expr.InsertExpr'.
 -}
-rawInsertExpr :: AnnotatedSqlMarshaller writeEntity readEntity -> Expr.InsertExpr -> Insert readEntity
-rawInsertExpr = Insert
+rawInsertExpr :: ReturningOption returningClause -> AnnotatedSqlMarshaller writeEntity readEntity -> Expr.InsertExpr -> Insert readEntity returningClause
```
Comment (qxjit) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Insert.hs:93]: It's probably worth documenting here that the caller is expected to pass a `ReturningOption` that accurately describes the `InsertExpr` that is passed.

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/ReturningOption.hs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{-# LANGUAGE GADTs #-}
+
+module Orville.PostgreSQL.Internal.ReturningOption (ReturningOption (..), ReturningClause, NoReturningClause) where
```
Comment (qxjit) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/ReturningOption.hs:3]: I'd prefer to have these export each on their own line

File: orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Expr/Update/UpdateExpr.hs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  )
+where
+
+import Data.Maybe (catMaybes)
+
+import Orville.PostgreSQL.Internal.Expr.Name (QualifiedTableName)
+import Orville.PostgreSQL.Internal.Expr.ReturningExpr (ReturningExpr)
+import Orville.PostgreSQL.Internal.Expr.Update.SetClauseList
+  ( SetClauseList,
+  )
```
Comment (qxjit) [inline orville-postgresql-libpq/src/Orville/PostgreSQL/Internal/Expr/Update/UpdateExpr.hs:20]: I'd prefer to have this import list on the line above with its module.

**Review by qxjit — APPROVED** [review]

**Pull request merged**